internal/proxy/proxy.go

```diff
@@ -51,34 +68,130 @@ func (proxy *DockerdProxy) addHandler(handler types.RequestHandler) {
 }
 
 // Start .
-func (proxy *DockerdProxy) Start(listener net.Listener) error {
-	addr := listener.Addr().String()
+func (proxy *DockerdProxy) Start(host ...Host) error {
+	switch len(host) {
+	case 0:
+		return errors.New("no listener is provided")
+	case 1:
+		return proxy.startOnHost(host[0])
+	default:
+		return proxy.startOnHosts(host)
+	}
+}
+
+// StartOnListeners .
+func (proxy *DockerdProxy) startOnHosts(hosts []Host) error {
+	errChannel := make(chan error)
+	for _, host := range hosts {
+		go func(host Host) {
+			errChannel <- proxy.startOnHost(host)
+		}(host)
+	}
+	err := <-errChannel
+	return err
+}
+
+func (proxy *DockerdProxy) startOnHost(host Host) error {
+	addr := host.Listener.Addr().String()
 	log.Infof("Starting proxy at %s", addr)
 	server := http.Server{
 		Addr:    addr,
 		Handler: proxy.mux,
 	}
-	return server.Serve(listener)
+	if host.Cert != "" {
+		return server.ServeTLS(host.Listener, host.Cert, host.Key)
+	}
+	return server.Serve(host.Listener)
 }
 
 func (proxy *DockerdProxy) dispatch(response http.ResponseWriter, request *http.Request) {
 	log.Infof("Incoming request, method = %s, url = %s", request.Method, request.URL.String())
+	if debug {
+		utils.PrintHeaders("ServerRequest", request.Header)
+	}
 
 	for _, handler := range proxy.handlers {
 		if handler.Handle(response, request) {
 			return
 		}
 	}
+	log.Info("handle other docker request, will forward stream")
 
 	var (
-		resp *http.Response
-		err  error
+		resp             *http.Response
+		err              error
+		protocolOverride bool
 	)
-	if resp, err = proxy.dockerSocket.Request(request); err != nil {
+	header := request.Header
+	header.Add("Host", request.Host)
+	// do a swallow copy of request
+	newRequest := *request
+	newRequest.Host = "docker"
+	protocolOverride = tryOverrideRequest(&newRequest)
+	if resp, err = proxy.dockerSocket.Request(&newRequest); err != nil {
 		log.Errorln("send request to docker socket error", err)
 		return
 	}
-	if err := proxy.netUtil.Forward(resp, response); err != nil {
-		log.Errorln("forward docker socket response error", err)
+	if resp.StatusCode != http.StatusSwitchingProtocols {
+		if debug {
+			log.Info("Type of body: ", reflect.TypeOf(resp.Body))
+		}
+		log.Info("Forward http response")
+		if err := utils.Forward(resp, response); err != nil {
+			log.Errorln("forward docker socket response error", err)
+		}
+	} else {
```

Comment:
the ```else``` block seems too long/nested to read, r they may refined?

Reply:
done